x/adminmodule/types/message_submit_proposal_legacy.go

```diff
@@ -0,0 +1,98 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+	govcodec "github.com/cosmos/cosmos-sdk/x/gov/codec"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	govtypesv1b1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
+	"github.com/cosmos/gogoproto/proto"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	_ sdk.Msg                          = &MsgSubmitProposalLegacy{}
+	_ cdctypes.UnpackInterfacesMessage = &MsgSubmitProposalLegacy{}
+)
+
+func NewMsgSubmitProposalLegacy(content govtypesv1b1.Content, proposer sdk.AccAddress) (*MsgSubmitProposalLegacy, error) {
+	m := &MsgSubmitProposalLegacy{
+		Proposer: proposer.String(),
+	}
+
+	err := m.SetContent(content)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *MsgSubmitProposalLegacy) GetContent() govtypesv1b1.Content {
+	content, ok := m.Content.GetCachedValue().(govtypesv1b1.Content)
+	if !ok {
+		return nil
+	}
+	return content
+}
+
+func (m *MsgSubmitProposalLegacy) SetContent(content govtypesv1b1.Content) error {
+	msg, ok := content.(proto.Message)
+	if !ok {
+		return fmt.Errorf("can't proto marshal %T", msg)
+	}
+	any, err := types.NewAnyWithValue(msg)
+	if err != nil {
+		return err
+	}
+	m.Content = any
+	return nil
+}
+
+func (m *MsgSubmitProposalLegacy) GetSigners() []sdk.AccAddress {
+	proposer, err := sdk.AccAddressFromBech32(m.Proposer)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{proposer}
+}
+
+func (m *MsgSubmitProposalLegacy) GetSignBytes() []byte {
+	bz := govcodec.ModuleCdc.MustMarshalJSON(m)
+	return sdk.MustSortJSON(bz)
+}
+
+// String implements the Stringer interface
+func (m *MsgSubmitProposalLegacy) String() string {
+	out, _ := yaml.Marshal(m)
+	return string(out)
+}
+
+// ValidateBasic implements Msg
+func (m *MsgSubmitProposalLegacy) ValidateBasic() error {
+	if m.Proposer == "" {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, m.Proposer)
```

Comment:
what's the point of using an empty address as a description? I'd rather put there something like "empty proposer field"

Reply:
This is the way how it's done in current admin module & `x/gov` in sdk. So there his no point to rewrite _legacy_ code

Reply:
anyway I fixed that c: